src/aws_signature_builder/mod.rs

```diff
@@ -0,0 +1,390 @@
+extern crate querystring;
+
+use crypto::digest::Digest;
+use crypto::sha2::Sha256;
+use hmac_sha256::HMAC;
+
+use hex;
+use chrono;
+
+use percent_encoding::{utf8_percent_encode, AsciiSet, CONTROLS};
+
+use std::collections::HashMap;
+
+/// https://url.spec.whatwg.org/#fragment-percent-encode-set
+const FRAGMENT: &AsciiSet = &CONTROLS.add(b' ').add(b'"').add(b'<').add(b'>').add(b'`');
+
+/// https://url.spec.whatwg.org/#path-percent-encode-set
+const PATH: &AsciiSet = &FRAGMENT.add(b'#').add(b'?').add(b'{').add(b'}');
+
+fn normalize_query_string(query: String) -> String {
+    let mut query_pairs = querystring::querify(&query);
+    query_pairs.sort_by(|a, b| a.0.cmp(&b.0));
+    return String::from(querystring::stringify(query_pairs).trim_end_matches("&"));
+}
+
+pub fn get_utc_aws_signature_datestrings() -> (String, String) {
+    let now = chrono::Utc::now();
+    let amzdate = now.format("%Y%m%dT%H%M%SZ").to_string();
+    let datestamp = now.format("%Y%m%d").to_string();
+    return (amzdate, datestamp)
+}
+
+pub fn generate_aws_signature_headers(
```

Comment:
This function is a beast!

Reply:
Haha, yeah.  This is a big insane pile of arguments.  It could definitely be improved.

I think this is related to [this comment above](https://github.com/sverch/aws-signature-proxy/pull/1#discussion_r337242110).  I could probably combine these arguments into various structs, like `AWSCredentials` and maybe even the full Request object along with it.

Reply:
Updated!  Now it only takes three arguments instead of 15.  :)